etc/systemd/system-generators/zfs-mount-generator.c

```diff
@@ -0,0 +1,1084 @@
+		    dataset, getpid(), mountfile, destdir, strerror(errno));
+		return (1 << 7);
+	}
+
+	fprintf(mountfile_f,
+	    OUTPUT_HEADER
+	    "[Unit]\n"
+	    "SourcePath=" FSLIST "/%s\n"
+	    "Documentation=man:zfs-mount-generator(8)\n"
+	    "\n"
+	    "Before=",
+	    cachefile);
+
+	if (p_systemd_before)
+		fprintf(mountfile_f, "%s ", p_systemd_before);
+	fprintf(mountfile_f, "zfs-mount.service");
+	if (requiredby)
+		fprintf(mountfile_f, " %s", requiredby);
+	if (wantedby && wantedby_append)
+		fprintf(mountfile_f, " %s", wantedby);
+
+	fprintf(mountfile_f,
+	    "\n"
+	    "After=");
+	if (p_systemd_after)
+		fprintf(mountfile_f, "%s ", p_systemd_after);
+	fprintf(mountfile_f, "%s\n", after);
+
+	fprintf(mountfile_f, "Wants=%s\n", wants);
+
+	if (bindsto)
+		fprintf(mountfile_f, "BindsTo=%s\n", bindsto);
+	if (p_systemd_requires)
+		fprintf(mountfile_f, "Requires=%s\n", p_systemd_requires);
+	if (p_systemd_requiresmountsfor)
+		fprintf(mountfile_f,
+		    "RequiresMountsFor=%s\n", p_systemd_requiresmountsfor);
+
+	fprintf(mountfile_f,
+	    "\n"
+	    "[Mount]\n"
+	    "Where=%s\n"
+	    "What=%s\n"
+	    "Type=zfs\n"
+	    "Options=defaults%s,zfsutil\n",
+	    p_mountpoint, dataset, opts);
+
+	(void) fclose(mountfile_f);
+
+	if (!requiredby && !wantedby)
+		return (0);
+
+	// Finally, create the appropriate dependencies
+	char *linktgt;
+	if (asprintf(&linktgt, "../%s", mountfile) == -1)
+		ABORT_ENOMEM();
+
+	char *dependencies[][2] = {
+		{"wants", wantedby},
+		{"requires", requiredby},
+		{}
+	};
+	for (__typeof__(&*dependencies) dep = &*dependencies; **dep; ++dep)
+		if ((*dep)[1]) {
+			for (char *reqby = strtok_r((*dep)[1], " ", &toktmp);
+			    reqby;
+			    reqby = strtok_r(NULL, " ", &toktmp)) {
+				char *depdir;
+				if (asprintf(
+				    &depdir, "%s.%s", reqby, (*dep)[0]) == -1)
+					ABORT_ENOMEM();
+
+				mkdirat(destdir_fd, depdir, 0755);
+				int depdir_fd = openat(destdir_fd, depdir,
+				    O_PATH | O_DIRECTORY | O_CLOEXEC);
+				if (depdir_fd < 0) {
+					fprintf(stderr, PROGNAME ": %s (%d): "
+					    "couldn't open %s under %s: %s\n",
+					    dataset, getpid(), depdir, destdir,
+					    strerror(errno));
+					free(depdir);
+					continue;
+				}
+
+				if (symlinkat(
+				    linktgt, depdir_fd, mountfile) == -1)
+					fprintf(stderr, PROGNAME ": %s (%d): "
+					    "couldn't symlink at "
+					    "%s under %s under %s: %s\n",
+					    dataset, getpid(), mountfile,
+					    depdir, destdir, strerror(errno));
+
+				(void) close(depdir_fd);
+				free(depdir);
+			}
+		}
+
+	return (0);
+}
+
+
+static int
+pool_enumerator(zpool_handle_t *pool, void *_)
+{
+	(void) _;
+	int ret = 0;
+
+	char *name = strdup(zpool_get_name(pool));
+	if (!name || !tsearch(name, &known_pools, STRCMP)) {
+		// This technically leaks if there's two pools
+		// with the same name.
+		// There aren't two pools with the same name.
+		free(name);
+		ret = ENOMEM;
+	}
+
+	zpool_close(pool);
+	return (ret);
+}
+
+int
```

Comment:
It looks like you are forking for each line (e.g. for each dataset). The semaphore stuff seems to be related to the noauto handling. Based on the comment above, I think you're blocking all the canmount=noauto children until you A) reach the end of the cache files) and B) all canmount=on children have completed. Then you let loose the canmount=noauto children. Did you consider a serial approach, where you process canmount=on as they occur, but defer canmount=noauto by building a list in memory, then processing that at the end?

Basically, what I'm wondering is: did you actually confirm that the parallel approach is faster, and sufficiently faster to justify the extra code complexity?

I'm _not_ saying it's not, since I have not evaluated this. So to be clear, I'm not asking you to go rewrite this serially.

Reply:
I mean, we do spend a non-immeasurable amount of time in fork(), I'm not disputing that, but each process dirties like single-digit shared pages, plus where-ever it moves its program break (usually not far).

But I did write it like this because (a) I started off with shell-version-but-it-`&`s-every-line, as seen in #11915, and (b) because it let me write each worker to be as fast as possible knowing I'm exiting ASAP, unmapping everything (and simplifies error handling a touch).

Plus, for me, just having a fence and a mutex was much easier and simpler to think about than achieving the same thing in a single process somehow. This may have been a different case if this weren't C, but.

That's not to say this doesn't leave *some* time on the table, but it, to me, is far outweighed by the clarity of this approach. If that time becomes significant or I'm deathly bored some time, I might revisit this to do everything in series.